From 5e349528423143f7318d12160679dc3831cd27c8 Mon Sep 17 00:00:00 2001
From: Patrick McCarty <patrick.mccarty@intel.com>
Date: Mon, 24 Jul 2017 16:52:56 -0700
Subject: [PATCH 3/6] Add a new 'pacdiscovery' (V2) service

On the Clear Linux OS for Intel Architecture, systemd-networkd is the
default network manager. Since it does not talk to pacrunner, we have
implemented a basic service called 'pacdiscovery' that performs a DNS
query for "wpad" and uses the query result to create pacrunner's proxy
configuration. To keep the configuration active, pacdiscovery keeps this
connection to pacrunner open, and the service sits idle until the system
reboots or the user manually stops or restarts it.

Note that the service is path activated through systemd whenever
systemd-networkd has set up a DHCP lease. So, static IP configurations
are not supported (at the moment) for auto-activation of pacdiscovery.

V2 of this patch changes significantly how pacdiscovery handles errors
and tries to be more robust in obtaining the WPAD address. It only
uses IPv4 since this is per spec, retries DNS lookups and will monitor
pacrunner and notice when it exits and restarts, and redo the
initialization sequence properly.

Signed-off-by: Patrick McCarty <patrick.mccarty@intel.com>
---
 Makefile.am              |  11 +-
 configure.ac             |   2 +
 src/discovery.c          | 287 +++++++++++++++++++++++++++++++++++++++
 src/pacdiscovery.conf    |   1 +
 src/pacdiscovery.path    |   8 ++
 src/pacdiscovery.service |  12 ++
 6 files changed, 320 insertions(+), 1 deletion(-)
 create mode 100644 src/discovery.c
 create mode 100644 src/pacdiscovery.conf
 create mode 100644 src/pacdiscovery.path
 create mode 100644 src/pacdiscovery.service

diff --git a/Makefile.am b/Makefile.am
index 452ba0a..9a87c22 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -66,7 +66,8 @@ endif
 
 duktape_libduktape_a_SHORTNAME = libduktape
 
-sbin_PROGRAMS = src/pacrunner
+sbin_PROGRAMS = src/pacrunner \
+		src/pacdiscovery
 
 src_pacrunner_SOURCES = $(gdbus_sources) $(builtin_sources) $(js_sources) \
 			src/main.c src/pacrunner.h \
@@ -75,6 +76,11 @@ src_pacrunner_SOURCES = $(gdbus_sources) $(builtin_sources) $(js_sources) \
 			src/client.c src/manager.c src/proxy.c \
 			src/manual.c src/js.h src/js.c
 
+src_pacdiscovery_SOURCES = src/discovery.c
+
+src_pacdiscovery_CFLAGS = @SYSTEMD_BUS_CFLAGS@
+src_pacdiscovery_LDADD = @SYSTEMD_BUS_LIBS@
+
 src_pacrunner_LDADD = $(builtin_libadd) @GLIB_LIBS@ @PTHREAD_LIBS@ \
 							@DBUS_LIBS@ -ldl
 
@@ -182,6 +188,9 @@ CLEANFILES = src/builtin.h
 
 EXTRA_DIST = src/genbuiltin src/pacrunner.conf $(test_scripts) \
 			src/pacrunner.service \
+			src/pacdiscovery.conf \
+			src/pacdiscovery.service \
+			src/pacdiscovery.path \
 			doc/architecture.txt doc/libproxy.txt \
 			doc/manager-api.txt doc/client-api.txt \
 			doc/manual-configuration.txt
diff --git a/configure.ac b/configure.ac
index 01f7d7a..961876c 100644
--- a/configure.ac
+++ b/configure.ac
@@ -92,6 +92,8 @@ if (test "${enable_v8}" = "yes"); then
 fi
 AM_CONDITIONAL(V8, test "${enable_v8}" = "yes")
 
+PKG_CHECK_MODULES([SYSTEMD_BUS],[libsystemd >= 221])
+
 PKG_CHECK_MODULES(GLIB, glib-2.0 >= 2.16, dummy=yes,
 				AC_MSG_ERROR(GLib >= 2.16 is required))
 AC_SUBST(GLIB_CFLAGS)
diff --git a/src/discovery.c b/src/discovery.c
new file mode 100644
index 0000000..8878ece
--- /dev/null
+++ b/src/discovery.c
@@ -0,0 +1,287 @@
+/*
+ * PAC file discovery service
+ *
+ * Copyright (C) 2013-2015 Intel Corporation
+ *
+ * This library is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+/*
+ * NOTE: the resolve() function is derived from the PACRunner jsengine code
+ * (copyright 2013, LGPL-2.1).
+ */
+
+#define _GNU_SOURCE
+#include <errno.h>
+#include <fcntl.h>
+#include <netdb.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <sys/socket.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <malloc.h>
+
+/* Certain static analysis tools do not understand GCC's __INCLUDE_LEVEL__
+ * macro; the conditional definition below is used to fix the build with
+ * systemd's _sd-common.h, included by the public systemd headers, which rely on
+ * the macro being defined and having an accurate value.
+ */
+#ifndef __INCLUDE_LEVEL__
+# define __INCLUDE_LEVEL__ 2
+#endif
+
+#include <systemd/sd-bus.h>
+#include <systemd/sd-event.h>
+
+static sd_event *event = NULL;
+static sd_bus *system_bus = NULL;
+
+char *resolve(char *str)
+{
+        struct addrinfo *info;
+        struct addrinfo hints;
+        char addr[65];
+        int err;
+
+        if (setenv("LC_ALL", "C", 1))
+                perror("Cannot set LC_ALL to C");
+
+        memset(&hints, 0, sizeof(struct addrinfo));
+        hints.ai_family = AF_INET;      /* IETF draft only required IPv4 */
+        hints.ai_socktype = SOCK_DGRAM; /* Datagram socket */
+        hints.ai_flags = 0;
+        hints.ai_protocol = 0;          /* Any protocol */
+
+        if ((err = getaddrinfo(str, NULL, &hints, &info)) != 0) {
+                fprintf(stderr, "failed getaddrinfo: %s\n", gai_strerror(err));
+                return NULL;
+        }
+
+        err = getnameinfo(info->ai_addr, info->ai_addrlen,
+                          addr, 65, NULL, 0, 0);
+        freeaddrinfo(info);
+        if (err < 0) {
+                fprintf(stderr, "failed getnameinfo: %s\n", gai_strerror(err));
+                return NULL;
+        }
+
+        return strdup(addr);
+}
+
+int send_to_pacrunner(char *host)
+{
+        sd_bus_error errstr = SD_BUS_ERROR_NULL;
+        sd_bus_message *obj = NULL;
+        char *configstr, *url;
+        int ret = -1;
+
+        // FIXME: need to support HTTPS URLs?
+        if (asprintf(&url, "http://%s/wpad.dat", host) < 0) {
+                return -1;
+        }
+
+        fprintf(stderr, "PAC url: %s\n", url);
+
+        ret = sd_bus_call_method(system_bus, "org.pacrunner",
+                                 "/org/pacrunner/manager",
+                                 "org.pacrunner.Manager",
+                                 "CreateProxyConfiguration",
+                                 &errstr, &obj, "a{sv}", 2,
+                                 "Method", "s", "auto",
+                                 "URL", "s", url);
+        if (ret < 0) {
+                fprintf(stderr, "Failed to create proxy config: %s\n",
+                        errstr.message);
+                goto bail;
+        }
+
+        if ((ret = sd_bus_message_read(obj, "o", &configstr)) < 0) {
+                fprintf(stderr, "Failed to read response from Pacrunner: %s\n",
+                        strerror(-ret));
+                goto bail;
+        }
+
+        fprintf(stderr, "Registered config: %s\n", configstr);
+
+        // FIXME: need to call DestroyProxyConfiguration when finished with the
+        // current configuration.
+
+        ret = 0;
+bail:
+        sd_bus_error_free(&errstr);
+        sd_bus_message_unref(obj);
+        free(url);
+        return ret;
+
+}
+
+static int state_fd = -1;
+static const char *state_file = "/run/pacrunner/pac_active";
+
+int create_state_file(void)
+{
+        int fd = -1;
+        int ret = -1;
+
+        if ((fd = open(state_file, O_WRONLY|O_CREAT, 00644)) < 0) {
+                fprintf(stderr, "Cannot create file %s: %s\n", state_file,
+                        strerror(errno));
+                goto fail;
+        }
+
+        state_fd = fd;
+
+        ret = 0;
+fail:
+        return ret;
+}
+
+static void remove_state_and_exit(int signum)
+{
+        /* state file may not have been created yet, so ignore errors */
+        unlink(state_file);
+        sd_event_exit(event, 0);
+}
+
+static int pacrunner_setup(void)
+{
+        char *domain = NULL;
+        int loop = 1;
+
+        /* don't give up too easily to resolve the wpad address */
+        for (;;) {
+                if ((domain = resolve("wpad")) != NULL) {
+                        break;
+                }
+
+                usleep(loop * 100000);
+
+                loop *= 2;
+                if (loop <= 32) {
+                        free(domain);
+                        continue;
+                }
+
+                fprintf(stderr, "Unable to find wpad host\n");
+                return 2;
+        }
+
+        if (send_to_pacrunner(domain) < 0) {
+                free(domain);
+                return 1;
+        }
+
+        free(domain);
+        return 0;
+}
+
+static int match_callback(sd_bus_message *m, void *userdata, sd_bus_error *ret_error)
+{
+        const char *s1, *s2, *s3; /* name, old_owner, new_owner */
+        int ret;
+
+        ret = sd_bus_message_read(m, "sss", &s1, &s2, &s3);
+        if (ret < 0) {
+                fprintf(stderr, "sd_bus_message_read: %s\n", strerror(-ret));
+                exit(EXIT_FAILURE);
+        }
+
+        if (strcmp(s1, "org.pacrunner") != 0) {
+                return 0;
+        }
+
+        if (strlen(s3) > 0) {
+                /* request main event loop to exit */
+                pacrunner_setup();
+        } else if (strlen(s2) > 0) {
+                /* pacrunner exited, wait a while and restart it */
+        }
+        return 0;
+}
+
+int main(void)
+{
+        int ret = 0;
+        struct sigaction sa;
+
+        memset(&sa, 0, sizeof(sa));
+        sa.sa_handler = remove_state_and_exit;
+        sigaction(SIGHUP, &sa, NULL);
+        sigaction(SIGINT, &sa, NULL);
+        sigaction(SIGTERM, &sa, NULL);
+
+        /* connect to the system bus */
+        ret = sd_bus_default_system(&system_bus);
+        if (ret < 0) {
+                fprintf(stderr, "Cannot connect to system bus: %s\n", strerror(-ret));
+                exit(EXIT_SUCCESS);
+        }
+
+        /* register to name owner changes in case pacrunner restarts */
+        ret = sd_bus_add_match(system_bus,
+                             NULL,
+                             "type='signal',interface='org.freedesktop.DBus',member='NameOwnerChanged'",
+                             match_callback,
+                             NULL);
+        if (ret < 0) {
+                fprintf(stderr, "Failed to add bus match\n");
+                ret = 3;
+                goto out;
+        }
+
+        /* create eventfd */
+        ret = sd_event_default(&event);
+        if (ret < 0) {
+                fprintf(stderr, "Failed to allocate event loop\n");
+                exit(EXIT_FAILURE);
+        }
+
+        /* and connect the bus so bus callbacks are processed */
+        ret = sd_bus_attach_event(system_bus, event, SD_EVENT_PRIORITY_NORMAL);
+        if (ret < 0) {
+                fprintf(stderr, "Failed to attach event\n");
+                exit(EXIT_FAILURE);
+        }
+
+        if (create_state_file() < 0) {
+                ret = 1;
+                goto out;
+        }
+
+        /* initial setup - once manually */
+        pacrunner_setup();
+
+        malloc_trim(0);
+
+        ret = sd_event_loop(event);
+        if (ret < 0) {
+                fprintf(stderr, "Failed event loop: %s\n", strerror(-ret));
+                ret = 2;
+                goto out;
+        }
+
+out:
+        sd_bus_unref(system_bus);
+        sd_event_unref(event);
+        if (state_fd >= 0) {
+                close(state_fd);
+        }
+        return ret;
+}
+
+/* vi: set ts=8 sw=8 sts=4 et tw=80 cino=(0: */
diff --git a/src/pacdiscovery.conf b/src/pacdiscovery.conf
new file mode 100644
index 0000000..1caf5bc
--- /dev/null
+++ b/src/pacdiscovery.conf
@@ -0,0 +1 @@
+d /run/pacrunner 0755 pacrunner pacrunner
diff --git a/src/pacdiscovery.path b/src/pacdiscovery.path
new file mode 100644
index 0000000..793971c
--- /dev/null
+++ b/src/pacdiscovery.path
@@ -0,0 +1,9 @@
+[Unit]
+Description=Proxy AutoConfig discovery service (lease directory watch)
+
+[Path]
+DirectoryNotEmpty=/run/systemd/netif/leases
+DirectoryNotEmpty=/run/NetworkManager/system-connections
+
+[Install]
+WantedBy=multi-user.target
diff --git a/src/pacdiscovery.service b/src/pacdiscovery.service
new file mode 100644
index 0000000..c2c35d5
--- /dev/null
+++ b/src/pacdiscovery.service
@@ -0,0 +1,13 @@
+[Unit]
+Description=Proxy AutoConfig discovery service
+Requires=systemd-resolved.service
+After=systemd-resolved.service
+
+[Service]
+ExecStart=/usr/sbin/pacdiscovery
+RuntimeDirectory=pacrunner
+SuccessExitStatus=2
+User=pacrunner
+
+[Install]
+WantedBy=multi-user.target
-- 
2.17.0

